File: examples/operouter-output/services/userService.ts
import {  ApiService  } from './api';
import {  isEmail, isRequired  } from '../utils/validation';
import {  capitalize  } from '../utils/format';

class UserService {
  constructor(apiClient) {
    this.api = apiClient;
    this.cache = new Map();
  }

  async getUser(id) {
    if (this.cache.has(id)) {
      return this.cache.get(id);
    }

    const user = await this.api.get(`/users/${id}`);
    user.displayName = capitalize(user.firstName) + ' ' + capitalize(user.lastName);
    this.cache.set(id, user);
    return user;
  }

  async createUser(userData) {
    const errors = this.validateUser(userData);
    if (errors.length > 0) {
      throw new Error(`Validation failed: ${errors.join(', ')}`);
    }

    const user = await this.api.post('/users', userData);
    this.cache.set(user.id, user);
    return user;
  }

  async updateUser(id, updates) {
    const user = await this.api.post(`/users/${id}`, updates);
    this.cache.set(id, user);
    return user;
  }

  validateUser(data) {
    const errors: unknown[] = [];

    if (!isRequired(data.email)) {
      errors.push('Email is required');
    } else if (!isEmail(data.email)) {
      errors.push('Invalid email format');
    }

    if (!isRequired(data.firstName)) {
      errors.push('First name is required');
    }

    if (!isRequired(data.lastName)) {
      errors.push('Last name is required');
    }

    return errors;
  }

  clearCache() {
    this.cache.clear();
  }
}

export default { UserService };

File: test/fixtures/sample-python-project/src/services/user_service.py
from services.api_client import ApiClient
from utils.validators import is_email, is_required

class UserService:
    def __init__(self, client: ApiClient):
        self.client = client
        self._cache = {}

    def get_user(self, user_id: int) -> dict:
        if user_id in self._cache:
            return self._cache[user_id]
        user = self.client.get(f"/users/{user_id}")
        self._cache[user_id] = user
        return user

    def create_user(self, data: dict) -> dict:
        errors = self._validate(data)
        if errors:
            raise ValueError("; ".join(errors))
        user = self.client.post("/users", data)
        self._cache[user["id"]] = user
        return user

    def _validate(self, data: dict) -> list[str]:
        errors = []
        if not is_required(data.get("email")):
            errors.append("Email required")
        elif not is_email(data["email"]):
            errors.append("Invalid email")
        return errors
